### PR TITLE
refactor: app-data-evaluator

### PR DESCRIPTION
**Pull request by chrismclarke**
PR Checklist

- [ ] PR title descriptive (can be used in release notes)

## Description
Work on PRs #2388 #2454 is currently blocked due to challenges trying to get item rows to evaluate correctly from items, data_items and external actions contexts.

At the core of the issue is the way in which a general template parser does a 1-time pass of all data before it is sent to the relevant items/data_items processors, replacing dynamic variables where known and leaving partial expressions as-is.

This becomes problematic when using items that need to evaluate some sort of `@item` logic, as associated variables could be outdated and/or replaced with values that no longer evaluator. E.g. `@item.name=== @local.name` will reach the expression as `@item.name===name_without_quotation`. Even if defining `@item.name==="@local.name"` we can run into issues with quoted `"undefined"` values and such. There's also additional issues created when working with item actions that target different items (e.g. `set_item: _index:1`). 

All of these issues are also extracted away behind multiple layers of templateParser logic which are incredibly confusing and hard to follow (lots of hack/temp workarounds to fix legacy bugs)

Moving forwards it seems like the best solution would be to completely detach the data_items processor from the main parser, and let it handle it's own evaluation methods. This also would offer a nice staging ground/proof-of-concept for implementing an alternative template parser (or at least offer ways to start decoupling some of the more complex parts)

This PR is a step towards decoupling, by refactoring the existing `AppStringParser` to better handle mixed use cases of JS and string evaluation. This parser will then be used as a drop-in replacement to process dynamic templates within data_items

This should unblock the immediate issues with #2454

## Dev Notes
Unfortunately I made the age-old mistake of both renaming a file and making changes, so the diffs appear larger than in reality. The main change is to change the order of priority of the previous `AppSringEvaluator` so that it first attempts to evaluate as JS before using a fallback of string replacement. This should allow for any valid javascript expression to be evaluated.

The updated tests should capture the main use-cases, including the previously broken and disabled test:
```
it("@row.first_name === 'Ada'", () => {
    pending("TODO - fix implementation to support");
    expect(evaluator.evaluate("@row.first_name === 'Ada'")).toEqual(true);
  });
 ``` 

## Git Issues

Closes #

## Screenshots/Videos

_If useful, provide screenshot or capture to highlight main changes_

### CODE REVIEW

**Review by jfmcquade — APPROVED** [review]
Looks good, it was helpful to discuss on the call last week

File: packages/shared/src/models/appDataEvaluator/appDataEvaluator.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+   * 2) Perform string replacement of individual variables
+   * @example
+   * ```js
+   * evaluateExpression(`Hello @row.first_name`)
+   * ```
+   * // output
+   * "Hello Ada"
+   */
+  private evaluateExpression(expression: string) {
+    try {
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply packages/shared/src/models/appDataEvaluator/appDataEvaluator.ts:100]: It's incredibly hacky, but we're working with the restriction that authors expect to be able to interchangeably write both valid javascript `@local.number > 3` and templated text `hello @local.name`. There was a time when we expected authors to wrap calculations in `@calc(....)` syntax, but that isn't used very frequently in some of the more modern examples. 

The other option would be to try and analyze in advance what type of expressions are being used, e.g. have scripts in the parser that look for any unquoted strings, however this also gets quite fiddly given that some text may still be valid javascript (e.g. `this.`, `Number(...)`, `true`. Long term this is likely better, but I'd say beyond the scope of this PR to investigate more deeply (I know there was an independent universal data parser RFC, would probably be the place to include)
